File: bitcamp-web-03/src/main/java/bitcamp/pms/dao/BoardDao.java
package bitcamp.pms.dao;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

import bitcamp.pms.domain.Board;

public class BoardDao {
    static {
        try {
            Class.forName("com.mysql.jdbc.Driver");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    String jdbcUrl;
    String username;
    String password;
    
    public BoardDao(String jdbcUrl, String username, String password){
        this.jdbcUrl = jdbcUrl;
        this.username = username;
        this.password = password;
    }
    public int delete(int no) throws Exception {
       
        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "delete from pms2_board where bno=?");) {
            
            stmt.setInt(1, no);
            return stmt.executeUpdate();
        } 
    }
    
    public List<Board> selectList() throws Exception {
        
        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "select bno,titl,cdt from pms2_board");
            ResultSet rs = stmt.executeQuery();) {
            
            ArrayList<Board> list = new ArrayList<>();
            while (rs.next()) {
                Board board = new Board();
                board.setNo(rs.getInt("bno"));
                board.setTitle(rs.getString("titl"));
                board.setCreatedDate(rs.getString("cdt"));
                list.add(board);
            }
            return list;
        }
    }

    public int insert(Board board) throws Exception {
      
        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "insert into pms2_board(titl,cont,cdt) values(?,?,now())");) {
            
            stmt.setString(1, board.getTitle());
            stmt.setString(2, board.getContent());
        
            return stmt.executeUpdate();
        }
    }

    public int update(Board board) throws Exception {
      
        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "update pms2_board set titl=?, cont=?, cdt=now() where bno=?");) {
            
            stmt.setString(1, board.getTitle());
            stmt.setString(2, board.getContent());
            stmt.setInt(3, board.getNo());
            return stmt.executeUpdate();
        }
    }

    public Board selectOne(int no) throws Exception {
        
        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "select bno,titl,cont,cdt from pms2_board where bno=?");) {
            
            stmt.setInt(1, no);
            
            try (ResultSet rs = stmt.executeQuery();) {
                if (!rs.next()) 
                    return null;
                
                Board board = new Board();
                board.setNo(rs.getInt("bno"));
                board.setTitle(rs.getString("titl"));
                board.setContent(rs.getString("cont"));
                board.setCreatedDate(rs.getString("cdt"));
                return board;
            }
        }  
    }
}

File: bitcamp-nodejs/src/step05/boarddao.js
// 주제: DAO 모듈 만들기

var pool;

exports.setConnectionPool = (connectionPool) => {
	pool = connectionPool;
};

exports.list = (pageNo = 1, pageSize = 3, handler) => {
	 var startIndex = (pageNo - 1) * pageSize;
	    pool.query('select bno, titl, cdt from pms2_board limit ?, ?',
	        [startIndex, pageSize],
	        function(err, results) {
	      handler(err, results);
	    });
};

exports.add = (data, handler) =>{
	 pool.query('INSERT INTO pms2_board(titl,cont,cdt) VALUES (?,?,now())',
	        	[data.title, data.content],
	            function(err, result) {
		 handler(err, result);
	        });
};

exports.update = (data, handler) =>{
	pool.query('update pms2_board set titl=?, cont=?, cdt=now() where bno=?', 
    		[data.title,
    			data.content, data.no],
    	     function(err, result) {
    	        handler(err, result);	   
    });	
};

exports.remove = (data, handler) =>{
	 pool.query('delete from pms2_board where bno=?', 
	     		[data.no],
	     	     function(err, result) {
	     	  handler(err, result);
	     });
};

exports.selectOne = (data, handler) =>{
	 pool.query('select bno, titl, cont, cdt from pms2_board where bno=?', 
	     		[data.no],
	     	     function(err, results) {
	     	  handler(err, results);
	     });
};